File: walk_detector.py
from activity_state import ActivityState
import math
import logging, sys
class Vector:
	import math
	def __init__(self, x = 0.0, y = 0.0, z = 0.0, timestamp = 0.0):
		self.x = x
		self.y = y
		self.z = z
		self.timestamp = timestamp

	def getMag(self):
		return math.sqrt(math.pow(self.x, 2) + math.pow(self.y, 2) + math.pow(self.z, 2))

	def greaterThanConst(self, const):
		if (self.x > const):
			return True
		if (self.y > const):
			return True
		if (self.z > const):
			return True
		return False;

	def equal(self, vec):
		if (self.x != vec.x):
			return False
		if (self.y != vec.y):
			return False
		if (self.z != self.z):
			return False
		return True

	def add(self, vec):
		result = Vector()
		result.x = self.x + vec.x
		result.y = self.y + vec.y
		result.z = self.z + vec.z
		return result

	def minus(self, vec):
		result = Vector()
		result.x = self.x - vec.x
		result.y = self.y - vec.y
		result.z = self.z - vec.z
		return result

	def mult(self, vec):
		result = Vector()
		result.x = self.x * vec.x
		result.y = self.y * vec.y
		result.z = self.z * vec.z
		return result

	def multConst(self, const):
		result = Vector()
		result.x = self.x * const
		result.y = self.y * const
		result.z = self.z * const
		return result

	def div(self, vec):
		result = Vector()
		if (vec.x == 0):
			result.x = 0
		else :
			result.x = self.x / vec.x
		if (vec.y == 0):
			result.y = 0
		else:
			result.y = self.y / vec.y
		if (vec.z == 0):
			result.z = 0
		else:
			result.z = self.z / vec.z
		return result 

	def divConst(self, const):
		if const == 0:
			raise ZeroDivisionError("Vec.divConst: Cannot divide by 0!")
		result = Vector()
		result.x = self.x / const
		result.y = self.y / const
		result.z = self.z / const
		return result

	def __str__(self):
		return str('X: %f Y: %f Z: %f' % (self.x, self.y, self.z))

class WalkDetector:
	gMin = 120
	gMax = 200
	gOpt = 0
	gAbsMin = 120
	gAbsMax = 240
	increment = 40

	logging.basicConfig(stream = sys.stderr, level = logging.DEBUG)

	def __init__(self, samples):
		# logging.info("Walk Detector")
		self.samples = list(map(lambda x: Vector(x[0], x[1], x[2], x[3]), samples))
		logging.info('Total samples: %d' % (len(self.samples)))


	def compute(self):
		state = ActivityState.IDLE
		numStepCtr = 0
		numSteps = 0
		logging.info("Begin computation")
		for i in range(len(self.samples)):
			highestCorrelation = self.getMaxCorrelation(i)
			if (self.getAccelStdDevOfMag(self.samples, i, self.gOpt) < 0.01) is True:
				state = ActivityState.IDLE
				logging.info("[IDLE STATE] timestamp: %d" % (self.samples[i].timestamp))
				numStepCtr = 0
			elif (highestCorrelation.greaterThanConst(0.7)) is True:
				state = ActivityState.WALKING
				self.gMin = self.gOpt - self.increment
				if (self.gMin < self.gAbsMin):
					self.gMin = self.gAbsMin
					self.gMax = self.gMin + (self.increment * 2)
				else:
					self.gMax = self.gOpt + self.increment
					if (self.gMax > self.gAbsMax):
						self.gMax = self.gAbsMax
						self.gMin = self.gMax - (self.increment * 2)
			if (state == ActivityState.IDLE):
				continue
			if (numStepCtr > self.gOpt / 2):
				numStepCtr = 0
				numSteps += 1
				logging.info("[WALK STATE] StepCount: %d Index:%d gOpt: %d" % (numSteps, i, self.gOpt))

			numStepCtr += 1

		return (ActivityState.IDLE,len(self.samples))

	def getMaxCorrelation(self, m):
		highestCorrelation = Vector()
		highestGamma = self.gMin
		for gamma in range(self.gMin, self.gMax):
			if (m + gamma + gamma) >= len(self.samples):
				break
			correlation = self.getAutoCorrelation(m, gamma)
			highestCorrelation = self.getHighestCorrelation(highestCorrelation, correlation)
			if (highestCorrelation.equal(correlation)):
				highestGamma = gamma
		self.gOpt = highestGamma
		return highestCorrelation

	def getAutoCorrelation(self, m, gamma):	
		meanVec = Vector()
		outputCorrelation = Vector()
		for k in range(gamma):
			if (m + k + gamma >= len(self.samples)):
				break;
			meanVec = self.getAccelMeanFromTill(m, gamma)
			left = self.samples[m + k].minus(meanVec)
			meanVec = self.getAccelMeanFromTill(m + gamma, gamma)
			right = self.samples[m + k + gamma].minus(meanVec)
			result = left.mult(right)
			outputCorrelation = outputCorrelation.add(result)
		
		stdDevLeftVec = self.getAccelStdDevFromTill(m, gamma)
		stdDevRightVec = self.getAccelStdDevFromTill(m + gamma, gamma)
		denominatorVec = stdDevLeftVec.mult(stdDevRightVec)
		denominatorVec = denominatorVec.multConst(gamma)
		outputCorrelation = outputCorrelation.div(denominatorVec)

		return outputCorrelation


	def getAccelStdDevFromTill(self, fromIdx, windowSize):
		newSampleDataSet = self.samples[fromIdx: fromIdx + windowSize]
		outputStdDevVec = self.getAccelStdDev(newSampleDataSet, windowSize)
		return outputStdDevVec

	def getAccelStdDev(self, samples, windowSize):
		meanVec = self.getAccelMean(samples, windowSize)
		stdDevVec = Vector()
		for i in range(windowSize):
			stdDevVec.x += math.pow(samples[i].x - meanVec.x, 2)
			stdDevVec.y += math.pow(samples[i].y - meanVec.y, 2)
			stdDevVec.z += math.pow(samples[i].z - meanVec.z, 2)
		stdDevVec.x = math.sqrt(stdDevVec.x / windowSize)
		stdDevVec.y = math.sqrt(stdDevVec.y / windowSize)
		stdDevVec.z = math.sqrt(stdDevVec.z / windowSize)
		return stdDevVec

	def getAccelMeanFromTill(self, fromIdx, windowSize):
		newSampleDataSet = []
		outputMeanVec = Vector()
		for k in range(windowSize):
			targetIdx = fromIdx + k
			newSampleDataSet.append(self.samples[targetIdx])
		outputMeanVec = self.getAccelMean(newSampleDataSet, windowSize)
		return outputMeanVec


	def getAccelMean(self, samples, windowSize):
		meanVec = Vector()
		for i in range(windowSize):
			meanVec = meanVec.add(samples[i])

		meanVec = meanVec.divConst(windowSize)
		return meanVec

	def getAccelStdDevOfMag(self, samples, fromIdx, windowSize):
		newSampleMagnitude = []
		for k in range(windowSize):
			targetIdx = fromIdx + k
			if (targetIdx >= len(samples)):
				break
			# logging.info('TargetIdx: %d k: %d fromIdx: %d' % (targetIdx, k, fromIdx))
			vecMag = samples[targetIdx].getMag()
			newSampleMagnitude.append(vecMag)
		stdDev = self.getStdDev(newSampleMagnitude)
		return stdDev

	def getStdDev(self, samples):
		mean = self.getMean(samples)
		stdDev = 0.0
		for i in range(len(samples)):
			stdDev += math.pow(samples[i]- mean, 2)

		stdDev = math.sqrt(stdDev / len(samples))
		return stdDev

	def getMean(self, samples):
		output = 0.0
		for i in range(len(samples)):
			output += samples[i]
		return output / len(samples)

	def getHighestCorrelation(self, left, right):
		leftSum = left.x + left.y + left.z
		rightSum = right.x + right.y + right.z
		if (leftSum > rightSum):
			return left
		return right

File: WalkDetector.java
import java.util.ArrayList;
import java.lang.Math;

public class WalkDetector {
private class Vector {
		public float x, y, z;
		public long timestamp;

		public Vector(float x, float y, float z) {
			this.x = x;
			this.y = y;
			this.z = z;

			this.timestamp = 0;
		}

		public Vector(ActivityData ad) {
			this.timestamp = ad.timestamp;

			this.x = ad.data.get(0);
			this.y = ad.data.get(1);
			this.z = ad.data.get(2);
		}

		public Vector add(Vector vec) {
			float outX, outY, outZ;
			outX = this.x + vec.x;
			outY = this.y + vec.y;
			outZ = this.z + vec.z;

			return new Vector(outX, outY, outZ);
		}

		public Vector minus(Vector vec) {
			float outX, outY, outZ;
			outX = this.x - vec.x;
			outY = this.y - vec.y;
			outZ = this.z - vec.z;

			return new Vector(outX, outY, outZ);
		}

		public Vector mult(Vector vec) {
			float outX, outY, outZ;
			outX = this.x * vec.x;
			outY = this.y * vec.y;
			outZ = this.z * vec.z;

			return new Vector(outX, outY, outZ);
		}
		
		public Vector multConst(float num) {
			float outX, outY, outZ;
			outX = this.x * num;
			outY = this.y * num;
			outZ = this.z * num;

			return new Vector(outX, outY, outZ);
		}
		
		public Vector div(Vector vec) {
			float outX, outY, outZ;
			outX = this.x / vec.x;
			outY = this.y / vec.y;
			outZ = this.z / vec.z;

			return new Vector(outX, outY, outZ);
		}

		public boolean equals(Vector vec) {
			if (x != vec.x) {
				return false;
			}
			if (y != vec.y) {
				return false;
			}
			if (z != vec.z) {
				return false;
			}
			return true;
		}
	
		public boolean moreThan(Float num) {
			if (x > num) {
				return true;
			}
			if (y > num) {
				return true;
			}
			if (z > num) {
				return true;
			}
			return false;
		}

		public float getMag() {
			return (float)Math.sqrt(Math.pow(x,2) + Math.pow(y,2) + Math.pow(z,2));
		}

		@Override
		public String toString() {
			return new String("X: " + this.x + " Y: " + this.y + " Z: " + this.z);
		}
	}

	private String state;
	private ArrayList<Float> samples;
	private ArrayList<Long> timestamps;
	private final int winSize = 20;
	private final float c = 1.0f/(2 * winSize + 1);
	private int sampleIndex = 0;
	private final float T1_THRESH = 2.0f;
	private final float T2_THRESH = 1.0f;
	private float moving_t1_thresh = T1_THRESH;
	private float moving_t2_thresh = T2_THRESH;
	private float b1, b2;
	private float prevB1;
	private final float MAX_ERROR_ALLOWED = 0.4f;
	private enum phase {
		STANCE,
		SWING
	}

	private phase curPhase;
	private long curTimeStamp;
	private boolean waitNextWindow;
	public WalkDetector() {
		state = ActivityState.IDLE;
		samples = new ArrayList<Float>();
		timestamps = new ArrayList<Long>();
		b1 = b2 = 0;
		curPhase = phase.STANCE;
		curTimeStamp = 0;
		waitNextWindow = false;
	}

	public OutputState compute(ArrayList<ActivityData> accelData) {
		waitNextWindow = false;
		for (ActivityData data : accelData) {
			Vector vec = new Vector(data).multConst(9.80665f);
			samples.add(vec.getMag());
			timestamps.add(data.timestamp);
			for (; sampleIndex < samples.size(); ++sampleIndex) {
				if (0 > sampleIndex - winSize) {
					continue;
				}
				if (sampleIndex + winSize >= samples.size()) {
					break;
				}
				float variance = getLocalVariance(sampleIndex);
				float stdDev = (float)Math.sqrt(variance);
				if (curPhase == phase.STANCE) {
					if (Math.abs(stdDev - moving_t1_thresh) <= MAX_ERROR_ALLOWED ) {
						moving_t1_thresh = 0.2f * moving_t1_thresh + stdDev * 0.8f;
						if (Math.abs(T1_THRESH - moving_t1_thresh) < 0.8f) {
							moving_t1_thresh = T1_THRESH * 0.7f + moving_t1_thresh * 0.3f;
						}
						System.out.println("Moving T1 thresh: " + moving_t1_thresh);
						b1 = moving_t1_thresh;
						curPhase = phase.SWING;
						state = ActivityState.WALK;
						curTimeStamp = getMeanTimestamp(sampleIndex);
					} else {
						b1 = 0;
					}
				}
				if (curPhase == phase.SWING) {
					if (stdDev < moving_t2_thresh) {
						b2 = moving_t2_thresh;
					}
					float forwardStd = lookForwardStd(sampleIndex);
					if (prevB1 < b1 && Math.abs(forwardStd - moving_t2_thresh) <= MAX_ERROR_ALLOWED) {
						moving_t2_thresh = 0.2f * moving_t2_thresh + forwardStd * 0.8f;
						if (Math.abs(moving_t2_thresh - T2_THRESH) < 0.8f) {
							moving_t2_thresh = T2_THRESH * 0.7f + moving_t2_thresh * 0.3f;
						}
						System.out.println("Moving T2 thresh: " + moving_t2_thresh);
						curPhase = phase.STANCE;
						state = ActivityState.IDLE;
						curTimeStamp = getMeanTimestamp(sampleIndex);
					}
				}
				if (waitNextWindow == true) {
					break;
				}
			}
		}
		return new OutputState(curTimeStamp , state);
	}

	private float getLocalVariance(int idx) {
		float variance = 0.0f;
		float mean = getLocalMean(idx);
		for (int i = idx - winSize; i < idx + winSize; ++i) {
			variance += Math.pow(this.samples.get(i) - mean, 2);
		}
		return variance * this.c;
	}

	private float getLocalMean(int idx) {
		float sum = 0.0f;
		for(int i = idx - winSize; i < idx + winSize; ++i) {
			sum += this.samples.get(i);
		}
		return sum * this.c;
	}

	private float lookForwardStd(int idx) {
		float maxStd = 0.0f; 
		for (int i = idx; i < idx + winSize; ++i) {
			if (i + winSize >= this.samples.size()) {
				waitNextWindow = true;
				sampleIndex = i - winSize;
				break;
			}
			float std = getLocalVariance(i);
			if (maxStd < std) {
				maxStd = std;
			}
		}
		return maxStd;
	}

	private long getMeanTimestamp(int idx) {
		long timestamp = 0;
		for (int i = idx - winSize; i < idx + winSize; ++i) {
			timestamp += timestamps.get(i);
		}
		return timestamp / (winSize * 2);
	}
}